### ISSUE TITLE
Invalid chars in string treated as equal in C# and unequal in F#

### ISSUE THREAD

**Issue by dipenpdev** [Microsoft/visualfsharp#5371]
Initialising a string in F# with explicitly invalid characters (e.g. for testing purposes) results in the invalid character literals being coerced to `\uFFFD`. Doing the same in C# preserves the invalid character literals.

This means (for example), that we can call `string.Normalize(..)` from C# with the string `"Foo\ude82\ude6eBar"` and raise an exception. but can't in F#.

#### Repro steps

Provide the steps required to reproduce the problem

1. Run this C# code
```c#
using System;

var fooBarInitFromString = "Foo\ude82\ude6eBar";
var fooBarInitFromCharArray = new String(new []{'F', 'o', 'o', '\ude82', '\ude6e', 'B', 'a', 'r'});

Console.WriteLine($"String version and Char version equal: '{(fooBarInitFromString == fooBarInitFromCharArray)}'");
Console.WriteLine($"String version contains replacement character: '{fooBarInitFromString.Contains("\uFFFD")}'");
```
Output:
```
String version and Char version equal: 'True'
String version contains replacement character: 'False'
```

2. Run this F# code
```f#
open System

let fooBarInitFromString = "Foo\ude82\ude6eBar"
let fooBarInitFromCharArray = new String([| 'F'; 'o'; 'o'; '\ude82'; '\ude6e'; 'B'; 'a'; 'r' |])
    
printfn "String version and Char version equal: '%b'" (fooBarInitFromString = fooBarInitFromCharArray)
printfn "String version contains replacement character: '%b'" (fooBarInitFromString.Contains("\uFFFD"))
```
Output:
```
String version and Char version equal: 'false'
String version contains replacement character: 'true'
```

#### Expected behavior

Same behaviour as C#, namely preservation of invalid character literals.

#### Actual behavior

Invalid character literals are coerced to '\uFFFD' in F# on string initialisation.

#### Known workarounds

Call string constructor from F# with a character array.

#### Related information 

Provide any related information 

IL generated by C#:
```
IL_0000: ldstr        bytearray (46 00 6f 00 6f 00 82 de 6e de 42 00 61 00 72 00 )
IL_0005: stloc.0      // fooBarInitFromString
```

IL generated by F#:
```
IL_0017: ldstr        "Foo��Bar"
IL_001c: ldstr        "String version"
IL_0021: call         void Program::normalize@9(string, string)
IL_0026: nop          
```

**Comment by zpodlovics** [comment]
You can use the verbatim string literal if you start your string with ```@``` according to the documentation: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/literals

```fsharp
> "Foo\ude82\ude6eBar".ToCharArray()|>Array.map byte;;                    
val it : byte [] = [|70uy; 111uy; 111uy; 130uy; 110uy; 66uy; 97uy; 114uy|]

> [| 'F'; 'o'; 'o'; '\ude82'; '\ude6e'; 'B'; 'a'; 'r' |] |> Array.map byte;;
val it : byte [] = [|70uy; 111uy; 111uy; 130uy; 110uy; 66uy; 97uy; 114uy|]

> @"Foo\ude82\ude6eBar".ToCharArray()|>Array.map byte;;
val it : byte [] =
  [|70uy; 111uy; 111uy; 92uy; 117uy; 100uy; 101uy; 56uy; 50uy; 92uy; 117uy;
    100uy; 101uy; 54uy; 101uy; 66uy; 97uy; 114uy|]

> let MyString1 = "Foo\ude82\ude6eBar";;               
val MyString1 : string = "Foo��Bar"

> let MyString2 = @"Foo\ude82\ude6eBar";;
val MyString2 : string = "Foo\ude82\ude6eBar"

> let MyString3 = @"Foo\ude82\ude6eBar"B;;
val MyString3 : byte [] =
  [|70uy; 111uy; 111uy; 92uy; 117uy; 100uy; 101uy; 56uy; 50uy; 92uy; 117uy;
    100uy; 101uy; 54uy; 101uy; 66uy; 97uy; 114uy|]


```
Unicode is surpringly difficult: https://github.com/Microsoft/visualfsharp/issues/4697#issuecomment-383604493

**Comment by dipenpdev** [comment]
Hmn, how does the verbatim string help? Take for example:

**[C#:](https://dotnetfiddle.net/OcKLIA)**
```c#
using System;
using System.Text;
					
"Foo\ude82\ude6eBar".Normalize(NormalizationForm.FormD);
```

**[F#:](https://dotnetfiddle.net/nLEMLs)**
```f#
open System
open System.Text
    
@"Foo\ude82\ude6eBar".Normalize(NormalizationForm.FormD)
```

The [C#:](https://dotnetfiddle.net/OcKLIA) version throws a `[System.ArgumentException: Invalid Unicode code point found at index 3.`, whereas the [F#:](https://dotnetfiddle.net/nLEMLs) version does not.

**Comment by zpodlovics** [comment]
As far as I know there are no direct ldstr + bytearray string initialization in F#, instead the string will be created with lot's of indirection. This example will throw an exception too. Take a look at the generated IL end, the byte sequence will be familiar.

```
.data cil I_00003500 = bytearray (
                 46 00 6F 00 6F 00 82 DE 6E DE 42 00 61 00 72 00) // F.o.o...n.B.a.r.
```

```fsharp
// Learn more about F# at http://fsharp.org

module Main

#nowarn "9"

open System
open System.Text

let b = [| 'F'; 'o'; 'o'; '\ude82'; '\ude6e'; 'B'; 'a'; 'r' |]
let s = System.String(b)

[<EntryPoint>]
let main argv =
    let s2 = s.Normalize(NormalizationForm.FormD);
    0 // return an integer exit code

```

```
//  Microsoft (R) .NET Framework IL Disassembler.  Version 4.5.22220.0



// Metadata version: v4.0.30319
.assembly extern System.Runtime
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:2:1:0
}
.assembly extern FSharp.Core
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:4:3:0
}
.assembly extern System.Diagnostics.Debug
{
  .publickeytoken = (B0 3F 5F 7F 11 D5 0A 3A )                         // .?_....:
  .ver 4:1:1:0
}
.assembly 'ldstr'
{
  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.FSharpInterfaceDataVersionAttribute::.ctor(int32,
                                                                                                      int32,
                                                                                                      int32) = ( 01 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 ) 
  .custom instance void [System.Runtime]System.Runtime.Versioning.TargetFrameworkAttribute::.ctor(string) = ( 01 00 18 2E 4E 45 54 43 6F 72 65 41 70 70 2C 56   // ....NETCoreApp,V
                                                                                                              65 72 73 69 6F 6E 3D 76 32 2E 31 01 00 54 0E 14   // ersion=v2.1..T..
                                                                                                              46 72 61 6D 65 77 6F 72 6B 44 69 73 70 6C 61 79   // FrameworkDisplay
                                                                                                              4E 61 6D 65 00 )                                  // Name.
  .custom instance void [System.Runtime]System.Reflection.AssemblyCompanyAttribute::.ctor(string) = ( 01 00 05 6C 64 73 74 72 00 00 )                   // ...ldstr..
  .custom instance void [System.Runtime]System.Reflection.AssemblyConfigurationAttribute::.ctor(string) = ( 01 00 07 52 65 6C 65 61 73 65 00 00 )             // ...Release..
  .custom instance void [System.Runtime]System.Reflection.AssemblyFileVersionAttribute::.ctor(string) = ( 01 00 07 31 2E 30 2E 30 2E 30 00 00 )             // ...1.0.0.0..
  .custom instance void [System.Runtime]System.Reflection.AssemblyInformationalVersionAttribute::.ctor(string) = ( 01 00 05 31 2E 30 2E 30 00 00 )                   // ...1.0.0..
  .custom instance void [System.Runtime]System.Reflection.AssemblyProductAttribute::.ctor(string) = ( 01 00 05 6C 64 73 74 72 00 00 )                   // ...ldstr..
  .custom instance void [System.Runtime]System.Reflection.AssemblyTitleAttribute::.ctor(string) = ( 01 00 05 6C 64 73 74 72 00 00 )                   // ...ldstr..
  .custom instance void [System.Runtime]System.Reflection.AssemblyVersionAttribute::.ctor(string) = ( 01 00 07 31 2E 30 2E 30 2E 30 00 00 )             // ...1.0.0.0..

  // --- The following custom attribute is added automatically, do not uncomment -------
  //  .custom instance void [System.Runtime]System.Diagnostics.DebuggableAttribute::.ctor(valuetype [System.Runtime]System.Diagnostics.DebuggableAttribute/DebuggingModes) = ( 01 00 03 00 00 00 00 00 ) 

  .hash algorithm 0x00008004
  .ver 1:0:0:0
}
.mresource public FSharpSignatureData.ldstr
{
  // Offset: 0x00000000 Length: 0x000008A6
}
.mresource public FSharpOptimizationData.ldstr
{
  // Offset: 0x000008B0 Length: 0x0000025C
}
.module ldstr.exe
// MVID: {5b573317-4eaf-e15b-a745-03831733575b}
.imagebase 0x00400000
.file alignment 0x00000200
.stackreserve 0x00100000
.subsystem 0x0003       // WINDOWS_CUI
.corflags 0x00000001    //  ILONLY
// Image base: 0x00007FEE95238000


// =============== CLASS MEMBERS DECLARATION ===================

.class public abstract auto ansi sealed Main
       extends [System.Runtime]System.Object
{
  .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 07 00 00 00 00 00 ) 
  .field static assembly valuetype '<PrivateImplementationDetails$ldstr>'/T1795_16Bytes@ field1796@ at I_00003500
  .custom instance void [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableState) = ( 01 00 00 00 00 00 00 00 ) 
  .method public specialname static char[] 
          get_b() cil managed
  {
    // Code size       6 (0x6)
    .maxstack  8
    IL_0000:  ldsfld     char[] '<StartupCode$ldstr>'.$Main::b@10
    IL_0005:  ret
  } // end of method Main::get_b

  .method public specialname static string 
          get_s() cil managed
  {
    // Code size       6 (0x6)
    .maxstack  8
    IL_0000:  ldsfld     string '<StartupCode$ldstr>'.$Main::s@11
    IL_0005:  ret
  } // end of method Main::get_s

  .method public static int32  main(string[] argv) cil managed
  {
    .entrypoint
    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.EntryPointAttribute::.ctor() = ( 01 00 00 00 ) 
    // Code size       26 (0x1a)
    .maxstack  4
    .locals init (string V_0)
    IL_0000:  ldc.i4.0
    IL_0001:  stsfld     int32 '<StartupCode$ldstr>'.$Main::init@
    IL_0006:  ldsfld     int32 '<StartupCode$ldstr>'.$Main::init@
    IL_000b:  pop
    IL_000c:  call       string Main::get_s()
    IL_0011:  ldc.i4.2
    IL_0012:  callvirt   instance string [System.Runtime]System.String::Normalize(valuetype [System.Runtime]System.Text.NormalizationForm)
    IL_0017:  stloc.0
    IL_0018:  ldc.i4.0
    IL_0019:  ret
  } // end of method Main::main

  .method private specialname rtspecialname static 
          void  .cctor() cil managed
  {
    // Code size       13 (0xd)
    .maxstack  8
    IL_0000:  ldc.i4.0
    IL_0001:  stsfld     int32 '<StartupCode$ldstr>'.$Main::init@
    IL_0006:  ldsfld     int32 '<StartupCode$ldstr>'.$Main::init@
    IL_000b:  pop
    IL_000c:  ret
  } // end of method Main::.cctor

  .property char[] b()
  {
    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 09 00 00 00 00 00 ) 
    .get char[] Main::get_b()
  } // end of property Main::b
  .property string s()
  {
    .custom instance void [FSharp.Core]Microsoft.FSharp.Core.CompilationMappingAttribute::.ctor(valuetype [FSharp.Core]Microsoft.FSharp.Core.SourceConstructFlags) = ( 01 00 09 00 00 00 00 00 ) 
    .get string Main::get_s()
  } // end of property Main::s
} // end of class Main

.class private abstract auto ansi sealed '<StartupCode$ldstr>'.$Main
       extends [System.Runtime]System.Object
{
  .field static assembly initonly char[] b@10
  .custom instance void [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableState) = ( 01 00 00 00 00 00 00 00 ) 
  .field static assembly initonly string s@11
  .custom instance void [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableState) = ( 01 00 00 00 00 00 00 00 ) 
  .field static assembly int32 init@
  .custom instance void [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [System.Diagnostics.Debug]System.Diagnostics.DebuggerBrowsableState) = ( 01 00 00 00 00 00 00 00 ) 
  .custom instance void [System.Runtime]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = ( 01 00 00 00 ) 
  .custom instance void [System.Diagnostics.Debug]System.Diagnostics.DebuggerNonUserCodeAttribute::.ctor() = ( 01 00 00 00 ) 
  .method private specialname rtspecialname static 
          void  .cctor() cil managed
  {
    // Code size       38 (0x26)
    .maxstack  8
    IL_0000:  ldc.i4.8
    IL_0001:  newarr     [System.Runtime]System.Char
    IL_0006:  dup
    IL_0007:  ldtoken    field valuetype '<PrivateImplementationDetails$ldstr>'/T1795_16Bytes@ Main::field1796@
    IL_000c:  call       void [System.Runtime]System.Runtime.CompilerServices.RuntimeHelpers::InitializeArray(class [System.Runtime]System.Array,
                                                                                                              valuetype [System.Runtime]System.RuntimeFieldHandle)
    IL_0011:  stsfld     char[] '<StartupCode$ldstr>'.$Main::b@10
    IL_0016:  call       char[] Main::get_b()
    IL_001b:  newobj     instance void [System.Runtime]System.String::.ctor(char[])
    IL_0020:  stsfld     string '<StartupCode$ldstr>'.$Main::s@11
    IL_0025:  ret
  } // end of method $Main::.cctor

} // end of class '<StartupCode$ldstr>'.$Main

.class private abstract auto ansi sealed '<StartupCode$ldstr>.$Ldstr'.AssemblyInfo
       extends [System.Runtime]System.Object
{
} // end of class '<StartupCode$ldstr>.$Ldstr'.AssemblyInfo

.class private abstract auto ansi sealed '<StartupCode$ldstr>.$.NETCoreApp,Version=v2.1'.AssemblyAttributes
       extends [System.Runtime]System.Object
{
} // end of class '<StartupCode$ldstr>.$.NETCoreApp,Version=v2.1'.AssemblyAttributes

.class private abstract auto ansi sealed beforefieldinit '<PrivateImplementationDetails$ldstr>'
       extends [System.Runtime]System.Object
{
  .class explicit ansi sealed nested assembly beforefieldinit T1795_16Bytes@
         extends [System.Runtime]System.ValueType
  {
    .pack 0
    .size 16
  } // end of class T1795_16Bytes@

} // end of class '<PrivateImplementationDetails$ldstr>'


// =============================================================

.data cil I_00003500 = bytearray (
                 46 00 6F 00 6F 00 82 DE 6E DE 42 00 61 00 72 00) // F.o.o...n.B.a.r.
// *********** DISASSEMBLY COMPLETE ***********************
```

**Comment by dipenpdev** [comment]
Yep, that's the workaround I've got at the moment too (in the issue desc) - I guess it's the best we've got. 
Just wondered whether it was something worth considering looking into.

Although a corner case, I feel the discrepancy in behaviour between C# and F# would come as a surprise to most.

**Comment by zpodlovics** [comment]
@dipenpdev You can always try to generate IL dynamically or try to use inline IL to create statically (but I really doubt it will work because the inline IL lexer/parser probably does not have ldstr and bytearray as keywords) something like I did it here: https://github.com/Microsoft/visualfsharp/issues/409#issuecomment-98714790

**Comment by abelbraaksma** [comment]
@dipenpdev, what version of F# are you using? The parsing issue was fixed a long time ago: #338, and when I copy-paste-run your code in FSI, I get this output (which is correct, but different from your output)

```
String version and Char version equal: 'true'
String version contains replacement character: 'false'
```

I also checked my codebase on string tests, where I use legal and illegal surrogate pairs in string literals using `\u` and while my code has a warning pointing to the above issue, it seems to be stable for a while.

However, my version of F# is a few weeks old (15.8 Preview 3), so perhaps this is a regression?

**Comment by zpodlovics** [comment]
What unicode version is supported in the F# source code? Unicode could be surpringly difficult:

"If the compiler is using Unicode 6.3 or higher (or a version earlier than 4.0) then U+180E is deemed to be in the Cf category, and is therefore valid within an identifier. In that case, it’s fine for it to be escaped as per the code above. At that point, the identifier in the second line of the method is deemed to be “the same as” stringx, so the output is b." [1]

```fsharp
open System

[<EntryPoint>]
let main argv =
    let stringx = "a"
    let string\u180ex = "b"
    System.Console.WriteLine(stringx)
    0 // return an integer exit code

```

```
error FS0010: Unexpected character '\' in pattern
```

[1] https://codeblog.jonskeet.uk/2014/12/01/when-is-an-identifier-not-an-identifier-attack-of-the-mongolian-vowel-separator/

**Comment by abelbraaksma** [comment]
> `let string\u180ex = "b"`

As far as I know you cannot use `\u` in literals in F#. But you _can_ use the U+180E in F# source code, and it is currently considered a valid character. Whether that is because of the yes/no `Cf` category or not, I don't know, so I tried another Cf-category character, which seems to be allowed by F#. This suggests that the accepted range of Unicode characters for use in identifiers is wider than C# (but I think that was a given, I just didn't know which chars are (dis)allowed).

The following script has all the proper, sometimes invisible characters. When copy/pasting, I don't know if it survives, you may need to create esp U+180E yourself, but you can use it in VS in the editor.

```f#
let stringx = "a"   // plain identifier
let string᠎x = "b"   // identifier with U+180E (Cf cat)
let string؂x = "c"  // identifier with U+0602 (Cf cat)
let t = "؂"
let value = System.Globalization.CharUnicodeInfo.GetUnicodeCategory(t, 0)
printfn "%s" stringx
printfn "%s" string᠎x
printfn "%s" string؂x
printfn "%A" value
```

This will print in FSI:

```f#
a
b
c
Format
val stringx : string = "a"
val string᠎x : string = "b"
val string؂x : string = "c"
val t : string = "؂"
val value : Globalization.UnicodeCategory = Format
val it : unit = ()
```

Bottom line: Cf characters are valid in F# identifiers, just not with `'\u` escaping.

But this issue was about string literals, and I think these work properly, at least on my system, so I wonder who the original issue was reproduced.

**Comment by dipenpdev** [comment]
@abelbraaksma, just updated my tooling and I get the correct behaviour on `15.7.20180730.4`. Think I was ~1 month behind, can't be sure as I didn't note the previous vsix version. 

What I did note was, before: 
`Visual F# Tools 10.1 for F# 4.1   15.7.0.0.  Commit Hash: 173513e369ffb7a1c4d5dccf83696d9aac2ab2d0`, 

After: 
`Visual F# Tools 10.1 for F# 4.1   15.7.0.0.  Commit Hash: 09dd2fa976239b6282f7118d97a2f510a4e72ded`,

I initially used dotnetfiddle to confirm/recreate the issue, I'm not sure what F# version it uses but it's still recreatable there: [C# Version](https://dotnetfiddle.net/7eZP6T), and [F# Version](https://dotnetfiddle.net/mptsSj).

**Comment by dipenpdev** [comment]
>However, my version of F# is a few weeks old (15.8 Preview 3), so perhaps this is a regression?

Not sure if this is a regression, but I think the above does confirm this isn't a compiler issue - not currently anyway.

This does seem broken in `FSharp.Core` though, which may explain the dotnetfiddle scenario. I span up a console application using `FSharp.Core 4.5.2` and targeting `.NET 4.6.1` (with the code in the issue description) - and the issue is present. Interestingly this appears broken as far back as `FSharp.Core 3.1.2` (and targeting `.NET 4`), which is the closest nuget package available to the date #348 was merged.

**Comment by abelbraaksma** [comment]
Yes, it was broken indeed, but it was fixed for quite a while. That you saw it in a recent build is worrisome, that suggests that there isn't a test covering this. @dsyme, if there isn't a test indeed, should we add a few to cover encoding issues?

**Comment by TIHan** [comment]
I'm not able to reproduce it either, even in 15.6.

I think it is absolutely a good idea to have a test case for this. I will gladly accept it.
